Split out remaining KO components

orangejenny opened the pull request:
## Technical Summary
Same idea as https://github.com/dimagi/commcare-hq/pull/35187, for the rest of our knockout components.

## Safety Assurance

### Safety story
There's some risk that I missed something. This isn't a high-risk PR, but it's probably a medium-risk one.

The impact would be a widget not getting instantiated, which would mean it would be missing from the page, although it should break the other js on the page. Those kind of bugs can be quite visible.

I did this by grepping for all usages of these widgets (`<inline-edit`, `<pagination`, `<search-box`, `<select-toggle`) plus the `inline_edit_trans` template tag and `SelectToggle` crispy widget, and then grepping for `hqwebapp/js/components.ko` to find everywhere that module is used as either a script tag or a js dependency, and cross-referencing those two lists.

I'm going to do some targeted testing on staging, primarily to convince myself that the places where the js dependency was *missing* are still working properly and don't indicate I'm misunderstanding something.

When reviewing, it's worth knowing that non-requirejs page (mostly app manager and reports) don't *need* to have the components added to javascript modules. Modules that are only used on non-requirejs pages don't need to specify their dependencies at all, but some of them do, and in those cases I updated the dependency lists.

On non-requirejs pages, all that matters is that the relevant script tag is included, which for these components it always is, since they're in [core_libraries](https://github.com/dimagi/commcare-hq/blob/b38e1c874a0f1b1645ee3da9dc30a4fef24c9660/corehq/apps/hqwebapp/templates/hqwebapp/includes/core_libraries.html#L111-L115).

### Automated test coverage

I don't think there are any tests that any of these components get properly instantiated.

### QA Plan

Not requesting QA. See safety story.

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

orangejenny committed:
Removed components.ko.js

orangejenny committed:
Added dependencies for small django apps

orangejenny committed:
Updated exports app

orangejenny committed:
Updated hqmedia

There are other places components are used in hqmedia, but they're on non-js-bundled pages
so they don't use the js dependencies.

orangejenny committed:
Updated app_manager

There are a bunch more components in app manager, but on non-js-bundled pages.

orangejenny committed:
Updated geospatial report

This module is really only being used in a non-js-bundled context, so it doens't need to
specify its dependencies, but as long is it's doing so, they might as well be the correct
dependencies.

orangejenny committed:
Updated reports

orangejenny committed:
Updated users app

orangejenny committed:
Updated data_interfaces

orangejenny committed:
Fixed misspelling

orangejenny commented:
Not fixing lint errors.

orangejenny committed:
Fixed some lint errors

orangejenny committed:
Bootstrap 5 Migration - Rebuilt diffs

mjriley commented on corehq/apps/hqwebapp/static/hqwebapp/js/components/inline_edit.js:
```diff
@@ -144,4 +144,17 @@ hqDefine('hqwebapp/js/components/inline_edit', [
+    ko.components.register('inline-edit', component);
+
+    $(function () {
+        _.each($('inline-edit'), function (el) {
+            var $el = $(el);
+            if ($el.data('apply-bindings') !== false) {
+                $(el).koApplyBindings();
+            }
+        });
```
I notice this block in each component, as it used to be in _components.ko.js_. Do you think it makes sense to create a common registration function? Specifically, the `$el.data('apply-bindings')` logic isn't that obvious, so I worry about the duplication.

orangejenny replied:
yeah, I don't feel great about this duplication either, I'll take a look at cleaning it up.

orangejenny replied:
Cleaned up in 61776dbfb0d

mjriley reviewed (commented):
Thanks for doing this, I love these changes! That said, non-blocking concern over the duplication in widget registration. Stopping short of approval over a question of whether we need a bridge for in-flight code. Is it possible that someone has in-flight code using components.ko.js and won't see a merge conflict with these changes, but will introduce errors? Should we leave a placeholder components.ko.js that imports all these individual components as a bridge?

orangejenny commented:
Noting that I tested all pages touched in this PR on staging to verify their expected widgets are shown.

orangejenny commented on corehq/apps/hqwebapp/static/hqwebapp/js/components/inline_edit.js:
```diff
@@ -144,4 +144,17 @@ hqDefine('hqwebapp/js/components/inline_edit', [
+    ko.components.register('inline-edit', component);
+
+    $(function () {
+        _.each($('inline-edit'), function (el) {
```
Point taken, although I'm going to ignore it. I have a vague opinion we should move away from both underscore and jquery in favor of Array functions, but we don't have official guidance around this. It'd be a good discussion topic.

orangejenny replied:
Ended up addressing this in 61776dbfb0d because it did seem silly to pull in an underscore dependency *just* for `_.each`

orangejenny committed:
DRYed up component registration

orangejenny commented:
@mjriley The question about bridging code is a good point. I'm inclined to just grep for `hqwebapp/js/components.ko` shortly before deploy to verify nothing new has been introduced. I don't think we have *that* much front end activity in the code base. Does that seem too sketchy?

orangejenny committed:
"Bootstrap 5 Migration - Rebuilt diffs"

gherceg commented on corehq/apps/registry/static/registry/js/registry_edit.js:
```diff
@@ -13,6 +13,8 @@ hqDefine("registry/js/registry_edit", [
     'hqwebapp/js/bootstrap5/knockout_bindings.ko', // openModal
     'hqwebapp/js/bootstrap5/main', // makeHqHelp
     'hqwebapp/js/multiselect_utils',
+    'hqwebapp/js/components/inline_edit',
+    'hqwebapp/js/components/pagination',
```
I don't know JS and am curious why these need to be included in here when components.ko wasn't included before?

mjriley commented:
> I'm inclined to just grep for `hqwebapp/js/components.ko` shortly before deploy to verify nothing new has been introduced. I don't think we have _that_ much front end activity in the code base. Does that seem too sketchy?

I think I'll defer to your judgment here? If we were to import them all back into `components.ko.js`, probably with a note that importing from that was deprecated, I guess we run the 'risk' of people used to the old version never changing their code and having no real reason to do so. Do we have local hosters that have made their own changes? I don't know how realistic that concern is. And hopefully current in-development branches check their changes on staging after this is merged?

mjriley reviewed (approved):
Changes look great, thanks

orangejenny committed:
Fixed data registry components

orangejenny committed:
Removed pagination dependency from registry_edit

This is only needed for the logs area, and it's already included in registry_logs.js

orangejenny commented:
@mjriley I'm not worried about localhosters having made code changes, just internal PRs. I'm comfortable with grepping for changes before deploy, I'll do that.

millerdev reviewed (approved):
Reviewed last two commits only.

orangejenny commented on corehq/apps/dashboard/static/dashboard/js/dashboard.js:
```diff
@@ -4,7 +4,7 @@ hqDefine("dashboard/js/dashboard", [
     'underscore',
     'hqwebapp/js/initial_page_data',
-    'hqwebapp/js/components.ko',    // pagination widget
+    'hqwebapp/js/components/pagination',
```
I'm pretty confident. The grepping described in the safety story should have found all instances of the widgets, and that list is what I used to add/update dependencies, not the comments.
